swiftide-indexing/src/loaders/file_loader.rs

```diff
@@ -5,118 +5,181 @@ use std::{
-fn file_has_extension(extensions: Option<&[impl AsRef<str>]>, path: &Path) -> bool {
-    extensions.as_ref().is_none_or(|exts| {
-        let Some(ext) = path.extension() else {
-            return false;
-        };
-        exts.iter()
-            .any(|e| e.as_ref() == ext.to_string_lossy().as_ref())
-    })
+/// An iterator that walks over the files in a directory and loads them.
+///
+/// This is a private struct that is used to implement the `FileLoader` iterator.
+struct Iter {
+    /// The walk instance that iterates over the files in the directory.
+    walk: Walk,
+    /// The extensions to include.
+    include_extensions: Option<Vec<String>>,
+    /// A span that tracks the current file loader.
+    span: Span,
+}
+
+impl Iterator for Iter {
+    type Item = anyhow::Result<Node>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let _span = self.span.enter();
+        loop {
+            // stop the iteration if there are no more entries
+            let Some(entry) = self.walk.next() else {
+                return None;
+            };
+
+            // propagate any errors that occur during the directory traversal
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => return Some(Err(err.into())),
+            };
+
+            if let Some(node) = self.load(entry) {
+                return Some(node);
+            }
+        }
+    }
+}
+
+impl Iter {
+    /// Creates a new `Iter` instance.
+    fn new(root: &Path, include_extensions: Option<Vec<String>>) -> Self {
+        let span = debug_span!("file_loader", root = %root.display());
+        tracing::debug!(parent: &span, extensions = ?include_extensions, "Loading files");
+        Self {
+            walk: Walk::new(root),
+            include_extensions,
+            span,
+        }
+    }
+
+    #[instrument(skip_all, fields(path = %entry.path().display()))]
+    fn load(&self, entry: DirEntry) -> Option<anyhow::Result<Node>> {
+        if entry.file_type().is_some_and(|ft| !ft.is_file()) {
+            // Skip directories and non-files
+            return None;
+        }
+        if let Some(extensions) = &self.include_extensions {
+            let Some(extension) = entry.path().extension() else {
+                tracing::trace!("Skipping file without extension");
+                return None;
+            };
+            let extension = extension.to_string_lossy();
+            if !extensions.into_iter().any(|ext| ext == &extension) {
+                tracing::trace!("Skipping file with extension {extension}");
+                return None;
+            }
+        }
+        tracing::debug!("Loading file");
+        match read_node(entry) {
+            Ok(node) => {
+                tracing::debug!(node_id = %node.id(), "Loaded file");
+                Some(Ok(node))
+            }
+            Err(err) => {
+                tracing::error!(error = %err, "Failed to load file");
+                Some(Err(err))
```

Comment:
Generally I avoid doing both log + return error, so perhaps this should be skipped. However looking at the current handling of the error, it doesn't seem like it currently gets an error log message, instead it's passed back to the caller. WDYT?

Reply:
Yeah, currently it's up to the pipeline implementer to handle, log or ignore it. So each failed read bubbles up in the pipeline. If you do not handle it, calling query on the pipeline will result in an error. There's a couple of helpers for that in the pipeline itself. Kwaak blatantly logs and ignores them.

With that, should we remove the log here? Or do you think we should accept a bit of duplication here to ease users into it?

Reply:
This is what it would look like if I added the following to my calling code to log pipeline errors:
```rust
    {
        Ok(_) => info!("Pipeline finished successfully"),
        Err(e) => {
            tracing::error!(error = ?e, "Pipeline failed");
            return Err(e);
        }
    }
```

<img width="1646" alt="image" src="https://github.com/user-attachments/assets/d6f3f0aa-7ee9-4a5a-bdd6-92b3a51b4478" />

The good thing about having the error logged here is that it carries the context of which loader / file has the problem.

I think perhaps leave the log message here, but keep it in mind for removal when thinking holistically about how error propagation works (i.e. when you eventually ditch anyhow and error types).